### PR TITLE
Early stopping

### LINKED ISSUE #25: Early stopping
Implement early stopping (early stopping of the training) if the log of the loss has not changed for threshhold number of epoch

This should be optional as we also will have the ability to save the state of the model and evaluate by loading the desired state

### PR DESCRIPTION
**Pull request by crangelsmith**
Fixes #25

- [x] If validation loss doesn't decrease in N number of epochs, early stopping training flag is true. This can be triggered by 'total_loss', 'reco_loss', 'affinity_loss' and 'kldiv_loss' or 'all' (configurable on input)
- [x] If early stopping trigger is true make all plots, save model and meta file.
- [x] If early stopping config flag and the internal flag is True, stop training, otherwise save checkpoints and continue training.
Todos:
- [x] Check that this is only evaluated when beta and gamma are on the nominal values in case of cyclical scheduling. 
    - [x] Make sure loss within patience is not within scheduling either
- [x] Add early stopping when loss is fluctuating within 1% of the value
- [x] Add early stopping when loss is diverging from training loss.
- [x] Only save early stopping state once, if training is ongoing. 
- [x] Add tests for different scenarios

Other minor changes:

- [x] On training only log at epoch level, instead of batch level
- [x] Addresses part of #220

### CODE REVIEW

**Comment by marjanfamili** [comment]
At the moment if I set early_stopping to false it still asks for minimum_train.

**Comment by crangelsmith** [comment]
Closing this PR as is no longer needed.
